apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,378 @@
+	// A reference to the placement that this update run is applied to.
+	// There can be multiple active update runs for each placement but
+	// it's up to the devOps to make sure they don't conflict with each other.
+	// +required
+	PlacementRef PlacementReference `json:"placementRef"`
+
+	// The resource snapshot index that this update run is based on.
+	// +required
+	ResourceSnapShotIndex string `json:"resourceSnapShotIndex"`
+
+	// The override snapshot index that this update run is based on.
+	// +required
+	OverrideSnapShotIndex string `json:"overrideSnapShotIndex"`
+
+	// The reference to the update strategy that specifies the stages and the sequence
+	// in which the application will be updated on the member clusters.
+	// +required
+	StagedUpdateStrategyRef StagedUpdateStrategyReference `json:"stagedRolloutStrategyRef"`
+}
+
+// PlacementReference is a reference to a placement object.
+type PlacementReference struct {
+	// Name is the name of the referenced StagedUpdateStrategy.
+	Name string `json:"name"`
+	// Namespace is the namespace where the referenced StagedUpdateStrategy is located.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// StagedUpdateStrategyReference is a reference to a StagedUpdateStrategy.
+type StagedUpdateStrategyReference struct {
+	// Name is the name of the referenced StagedUpdateStrategy.
+	Name string `json:"name"`
+	// Namespace is the namespace where the referenced StagedUpdateStrategy is located.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// +genclient
+// +genclient:namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StagedUpdateStrategy defines a reusable strategy that specifies the stages and the sequence
+// in which the application will be updated on the member clusters.
+type StagedUpdateStrategy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Stage configurations for each update stage.
+	// +required
+	Stages []StageConfig `json:"stages"`
+
+	// TODO: Add alerting configuration.
+}
+
+// StagedUpdateStrategyList contains a list of StagedUpdateStrategy.
+// +kubebuilder:resource:scope="Namespaced"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type StagedUpdateStrategyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []StagedUpdateStrategy `json:"items"`
+}
+
+// StageConfig describes a single update stage group configuration.
+// The clusters in each stage are updated sequentially for now.
+// We will stop the update if any of the updates fail.
+type StageConfig struct {
+	// The name of the stage. This MUST be unique within the same StagedUpdateStrategy.
+	// +required
+	Name string `json:"name"`
```

Comment:
not at the moment since the rest of the fields don't really apply to that stage. We were thinking of mix the to be deleted clusters with the new/updated clusters but want to keep it simple for now.